Guarded insertion into filtered_arrivals_ in sta::Search

jbylicki opened the pull request:
This PR fixes a crash occurring when multiple threads (in testing 4+ threads resulted in a high chance of this crash happening) concurrently wrote to filtered_arrivals_ in Search::setVertexArrivals. An example stack trace of the crash is:

```
 0# 0x000058A238E94DE0 in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
 1# 0x00007D00A3FC8D60 in /lib/x86_64-linux-gnu/libc.so.6
 2# sta::Vertex::objectIdx() const in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
 3# sta::ObjectTable<sta::Vertex>::objectId(sta::Vertex const*) in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
 4# sta::Graph::id(sta::Vertex const*) const in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
 5# sta::VertexIdLess::operator()(sta::Vertex const*, sta::Vertex const*) const in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
 6# std::_Rb_tree<sta::Vertex*, sta::Vertex*, std::_Identity<sta::Vertex*>, sta::VertexIdLess, std::allocator<sta::Vertex*> >::_M_get_insert_unique_pos(sta::Vertex* const&) in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
 7# std::pair<std::_Rb_tree_iterator<sta::Vertex*>, bool> std::_Rb_tree<sta::Vertex*, sta::Vertex*, std::_Identity<sta::Vertex*>, sta::VertexIdLess, std::allocator<sta::Vertex*> >::_M_insert_unique<sta::Vertex* const&>(sta::Vertex* const&) in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
 8# std::set<sta::Vertex*, sta::VertexIdLess, std::allocator<sta::Vertex*> >::insert(sta::Vertex* const&) in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
 9# sta::Search::setVertexArrivals(sta::Vertex*, sta::TagGroupBldr*) in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
10# sta::ArrivalVisitor::visit(sta::Vertex*) in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
11# 0x000058A2393603A0 in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
12# 0x000058A2393618D2 in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
13# 0x000058A2393617B4 in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
14# 0x000058A239361603 in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
15# std::function<void (int)>::operator()(int) const in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
16# sta::DispatchQueue::dispatch_thread_handler(unsigned long) in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
17# void std::__invoke_impl<void, void (sta::DispatchQueue::*)(unsigned long), sta::DispatchQueue*, unsigned long>(std::__invoke_memfun_deref, void (sta::DispatchQueue::*&&)(unsigned long), sta::DispatchQueue*&&, unsigned long&&) in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
18# std::__invoke_result<void (sta::DispatchQueue::*)(unsigned long), sta::DispatchQueue*, unsigned long>::type std::__invoke<void (sta::DispatchQueue::*)(unsigned long), sta::DispatchQueue*, unsigned long>(void (sta::DispatchQueue::*&&)(unsigned long), sta::DispatchQueue*&&, unsigned long&&) in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
19# void std::thread::_Invoker<std::tuple<void (sta::DispatchQueue::*)(unsigned long), sta::DispatchQueue*, unsigned long> >::_M_invoke<0ul, 1ul, 2ul>(std::_Index_tuple<0ul, 1ul, 2ul>) in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
20# std::thread::_Invoker<std::tuple<void (sta::DispatchQueue::*)(unsigned long), sta::DispatchQueue*, unsigned long> >::operator()() in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
21# std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (sta::DispatchQueue::*)(unsigned long), sta::DispatchQueue*, unsigned long> > >::_M_run() in /data/OpenROAD-flow-scripts/tools/install/OpenROAD/bin/openroad
22# 0x00007D00A4376ED0 in /usr/lib/x86_64-linux-gnu/libstdc++.so.6
23# 0x00007D00A69EAEA7 in /lib/x86_64-linux-gnu/libpthread.so.0
24# clone in /lib/x86_64-linux-gnu/libc.so.6
```

The insertions have been guarded with a mutex to prevent concurrent writing

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/parallaxsw/OpenSTA?pullRequest=35) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/parallaxsw/OpenSTA?pullRequest=35) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/parallaxsw/OpenSTA?pullRequest=35) it.</sub>

parallaxsw commented:
Using std::lock_guard instead of std::unique_lock is a good idea so I changed the existing locks.
I left this one for your PR, but I do ask that you use the typedef in Mutex.hh and name the variable
_lock_ instead of _mutex_ to be consistent with all of the other uses of mutex.

jbylicki commented:
> Using std::lock_guard instead of std::unique_lock is a good idea so I changed the existing locks. I left this one for your PR, but I do ask that you use the typedef in Mutex.hh and name the variable _lock_ instead of _mutex_ to be consistent with all of the other uses of mutex.

Both requested changes are now implemented.

parallaxsw commented on include/sta/Search.hh:
```diff
@@ -30,6 +30,7 @@
 #include "SearchClass.hh"
 #include "SearchPred.hh"
 #include "VertexVisitor.hh"
+#include "Mutex.hh"
```
This include is unnecessary

parallaxsw commented on include/sta/Search.hh:
```diff
@@ -606,6 +607,7 @@ protected:
   ExceptionFrom *filter_from_;
   ExceptionTo *filter_to_;
   VertexSet *filtered_arrivals_;
+  UniqueLock filtered_arrivals_lock_;
```
This should be std::mutex
See any of the other lock variables as an example.
UniqueLock is for locking the mutex.

jbylicki replied:
I presume then the declaration should be then `std::mutex filtered_arrivals_lock_;` in `Search.hh`, and the lock have the form of `std::lock_guard<sta::UniqueLock> guard(this->filtered_arrivals_lock_);`.

Unless I have assumed wrong then, this will produce a type error on `lock_guard` - the type of `lock_guard<T_>` should be the same as of the lock that is being guarded. I can change the implementation to `UniqueLock lock(this->filtered_arrivals_lock_);` but I understood from the first message that `lock_guard` is more desirable.

parallaxsw replied:
There are 6 mutex locks defined in Search.hh.
Make this lock definition and use in Search.cc the same as all of the others.
